### PR TITLE
overlord/devicestate: set new state after reprovision

### PR DESCRIPTION
**Pull request by valentindavid**

### CODE REVIEW

File: secboot/secboot_sb_test.go
Comment (ZeyadYasser) [inline secboot/secboot_sb_test.go:1]: tests needs to be extended to check for the returned serialized pcr profile

File: overlord/fdestate/fdestate_test.go [hunk 1]
```diff
@@ -1076,3 +1079,51 @@ func (s *fdeMgrSuite) TestExpandSystemContainerRoles(c *C) {
...
+
+func (s *fdeMgrSuite) TestStateUpdate(c *C) {
+	st, err := fdestate.InitialState([]byte("primary"))
+	c.Assert(err, IsNil)
+
+	models := []secboot.ModelForSealing{
+		&mockModel{},
+	}
+
+	st.UpdateParameters("run+recover", "container-role", []string{"run"}, models, secboot.SerializedPCRProfile(`"serialized-profile"`))
```
Comment (ZeyadYasser) [inline overlord/fdestate/fdestate_test.go]: Please extend or add test to also check `UpdatePCRHandle`.

File: overlord/fdestate/fdestate_test.go [hunk 2]
```diff
@@ -1076,3 +1079,51 @@ func (s *fdeMgrSuite) TestExpandSystemContainerRoles(c *C) {
...
+	c.Assert(err, IsNil)
+	runRecoverRole, hasRunRecoverRole := st.KeyslotRoles["run+recover"]
+	c.Assert(hasRunRecoverRole, Equals, true)
+	containerRole, hasContainerRole := runRecoverRole.Parameters["container-role"]
+	c.Assert(hasContainerRole, Equals, true)
+
+	c.Assert(containerRole.Models, HasLen, 1)
+	c.Check(containerRole.Models[0].Model(), Equals, "mock-model")
+	c.Check(containerRole.BootModes, DeepEquals, []string{"run"})
+	c.Check(containerRole.TPM2PCRProfile, DeepEquals, secboot.SerializedPCRProfile(`"serialized-profile"`))
```
Comment (ZeyadYasser) [inline overlord/fdestate/fdestate_test.go:1132]: Maybe add a check that PCR handle was not changed by `UpdateParameters`

File: boot/makebootable_test.go [hunk 3]
```diff
@@ -2769,7 +2770,7 @@ func (s *makeBootable20Suite) TestMakeSystemRunnableReprovision(c *C) {
 
 	var protector secboot.KeyProtectorFactory
 
-	err = boot.MakeRunnableSystemReprovision(model, protector, encryptionSetup)
+	err = boot.MakeRunnableSystemReprovision(model, protector, encryptionSetup, nil)
```
Comment (ZeyadYasser) [inline boot/makebootable_test.go]: we are missing a test variant where initial state is non-nil, just to make sure it gets passed down to SealKeyForBootChains

File: overlord/devicestate/handlers_reprovision.go [hunk 4]
```diff
@@ -413,23 +413,32 @@ func (m *DeviceManager) doReprovision(t *state.Task, _ *tomb.Tomb) error {
...
 		keyProtector,
 		encryptionParams,
+		&fdeState,
 	)
 
 	if err != nil {
 		return fmt.Errorf("cannot make system runnable: %v", err)
 	}
 
 	// Step 7. Swap the state
```
Comment (ZeyadYasser) [inline overlord/devicestate/handlers_reprovision.go]: should this now b e step 6 and the below step 7 to avoid confusion?

File: overlord/fdestate/backend/seal.go [hunk 5]
```diff
@@ -198,8 +226,12 @@ func sealKeyForBootChainsHook(method device.SealingMethod, key, saveKey secboot.
 		sealingParams.AuxKeyFile = filepath.Join(boot.InstallHostFDESaveDir, "aux-key")
 	}
 
+	var models []secboot.ModelForSealing
 	for _, runChain := range params.RunModeBootChains {
 		sealingParams.Model = runChain.ModelForSealing()
+		models = append(models, runChain.ModelForSealing())
+		// We assume that factory-reset/installation/reprovision always reseal with one unique model.
+		// FIMXE: for reprovision, we might need to actually clean up the modeenv.
 		break
```
Comment (ZeyadYasser) [inline overlord/fdestate/backend/seal.go:250]: Why was this assumption and the corresponding `break` removed from the other sealing method but not for hooks based sealing? is there a reason for not having both be consistent?

**Review by ZeyadYasser — COMMENTED** [review]
Looks good, Thanks! I left some comments mostly related to testing.

File: overlord/devicestate/handlers_reprovision.go (hunk 4 above)
Comment (valentindavid) [reply overlord/devicestate/handlers_reprovision.go]: We actually swap the state after. This is just preparation for it. I will move the comment.

File: overlord/fdestate/backend/seal.go (hunk 5 above)
Comment (valentindavid) [reply overlord/fdestate/backend/seal.go:250]: That was easy to remove the other one. But here we have `sealingParams.Model` that can only contain one model. So we do need to refactor `secboot.SealKeysWithProtector` to support that.

I wonder if I should just make reprovision ensure there is only one for now and fail.

Comment (valentindavid) [reply overlord/fdestate/backend/seal.go:250]: I will prepare a follow up PR. I think I need to do that refactor.

Comment (valentindavid) [reply overlord/fdestate/backend/seal.go:250]: https://github.com/canonical/snapd/pull/17474/changes/3cace4ab2befc732097284c179d6661158290b26

File: tests/nested/manual/core-recover-from-recovery/task.yaml [hunk 6]
```diff
@@ -123,8 +123,40 @@ execute: |
...
+
+      remote.exec "sudo snap debug api /v2/systems?running=true 2>/dev/null" | gojq '.result["storage-encryption"].support' | MATCH "available"
+
+      echo '{"action":"generate-recovery-key"}' | remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems" >rkey-reprovision.resp
+      gojq --raw-output '.result."recovery-key"' < rkey-reprovision.resp > rkey-reprovision.out
+
+      if os.query is-ubuntu-le 24.04; then
+        echo "Reprovision is not expected to work on UC24 unless installed with tokens. Skipping." 1>&2
+        exit 0
+      fi
```
Comment (ZeyadYasser) [inline tests/nested/manual/core-recover-from-recovery/task.yaml:139]: This means the this part of the test will never run as the only spread system is 24.04, no?

**Review by ZeyadYasser — COMMENTED** [review]
Overall LGTM, Thank you! left a comment on a likely stale spread test `‎tests/nested/manual/core-recover-from-recovery`.

File: tests/nested/manual/core-recover-from-recovery/task.yaml (hunk 6 above)
Comment (valentindavid) [reply tests/nested/manual/core-recover-from-recovery/task.yaml:139]: If it is "less or equal" to 24.04, it will `exit 0` and skip the reprovision.

Comment (valentindavid) [reply tests/nested/manual/core-recover-from-recovery/task.yaml:139]: So for now it will only run on 26.04.

Comment (valentindavid) [reply tests/nested/manual/core-recover-from-recovery/task.yaml:139]: Ah wait... I missed the "systems".

Comment (valentindavid) [reply tests/nested/manual/core-recover-from-recovery/task.yaml:139]: Fixed it.

**Review by ZeyadYasser — APPROVED** [review]
LGTM, Thank you!

File: boot/seal.go [hunk 7]
```diff
@@ -183,13 +184,24 @@ type SealKeyForBootChainsParams struct {
 	KeyProtectorFactory secboot.KeyProtectorFactory
 }
 
+// InitialFDEState is a FDE state being built during initial sealing.
+type InitialFDEState interface {
```
Comment (pedronis) [inline boot/seal.go]: this is a bit of a odd place for this, at this point my preference would probably be to define this type and explicit  SealKeyForBootChainFunc and ResealKeyForBootChainFunc types in gadget/device

that is also not great but better than here

The other approach is just to use an empty interface instead as we are not calling those methods from here

**Review by pedronis — COMMENTED** [review]
code org nitpick

File: boot/seal.go (hunk 7 above)
Comment (valentindavid) [reply boot/seal.go]: You want the function types to be defined as SealKeyForBootChainFunc and ResealKeyForBootChainFunc in gadget/device? They depend on a lot of types. That would need moving a lot of definitions there.

Comment (pedronis) [reply boot/seal.go]: then lets use an empty marker interface here like

type InitialSealState interface{}

and do some casting with internal errors in fdestate

File: overlord/fdestate/backend/seal.go [hunk 8]
```diff
@@ -177,15 +202,28 @@ func sealFallbackObjectKeys(
...
 	// can be used if ubuntu-data and ubuntu-boot are corrupted or unavailable.
-
-	if _, err := secbootSealKeys(fallbackKeySealRequests(key, saveKey, factoryResetKeyPath, useTokens), sealKeyParams); err != nil {
+	_, pcrProfile, err := secbootSealKeys(fallbackKeySealRequests(key, saveKey, factoryResetKeyPath, useTokens), sealKeyParams)
+	if err != nil {
 		return fmt.Errorf("cannot seal the fallback encryption keys: %v", err)
 	}
 
+	if fdeState != nil {
+		// FIXME: we should create different pcr profile for ech.
```
Comment (pedronis) [inline overlord/fdestate/backend/seal.go]: what is this about?

**Review by pedronis — COMMENTED** [review]
question about one of the FIXMEs

File: overlord/fdestate/backend/seal.go (hunk 8 above)
Comment (valentindavid) [reply overlord/fdestate/backend/seal.go]: Fixed it.

I realized that I forgot to split the pcr profile for save a data when doing the FDE refactoring. Data should not unlock when factory-reset is used in the kernel command line.

Comment (valentindavid) [reply overlord/fdestate/backend/seal.go]: I have created SNAPDENG-37393 for that.

**Review by pedronis — APPROVED** [review]
+1 but we should think whether it's ok to land this partially for 2.78 or we need to wait for it to be fully fixed before it is landed

**Comment by valentindavid** [comment]
> +1 but we should think whether it's ok to land this partially for 2.78 or we need to wait for it to be fully fixed before it is landed

I think we can land. We need to tell desktop that if enroll passphrase/pin fails after reprovision, they should recommend a reboot.

We already landed the status. So we should just land all we can now.
